python/ray/train/callbacks/results_preprocessors/aggregate.py

```diff
@@ -0,0 +1,136 @@
+import logging
+from typing import Any, Callable, Dict, List, Tuple, Optional, Set, Union
+
+import numpy as np
+
+from ray.train.callbacks.results_preprocessors.preprocessor import ResultsPreprocessor
+from ray.train.constants import (
+    TIME_THIS_ITER_S,
+    TRAINING_ITERATION,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class AggregateResultsPreprocessor(ResultsPreprocessor):
+    """A preprocessor that aggregates training metrics from all workers.
+
+    Args:
+        aggregators (Optional[Dict[str, Optional[Tuple|Callable[[List], Any]]]]):
+            A Dict of aggregation methods. The key is the name of the aggregated
+            metric. The value is the aggregation method. If it is a function, the
+            aggregation will be performed by calling this function. If the value
+            is a tuple (var1, var2), var1 specifies which metrics to be aggregated,
+            and var2 is the averaging weights. When var2 is None, the weight will
+            be taken to be equal for all workers. var1 and var2 are expected to be
+            reported using `train.report()`.
```

Comment:
This is good for the average and weighted average case, specifically for a small number of metrics, though I'm wondering how to extend this so the user can easily grab the aggregated (mean, max, min, std, ...) of all their metrics in a few lines of code. 

I roughly put together some pseudo-code for this:

```python
VALID_AGGREGATE_TYPE = Union[int,
        float,
        np.float32,
        np.float64,
        np.int32,
        np.int64
]

#########################################################
# Aggregation Functions
#########################################################

class AggregateFn():
    def aggregate():
        raise NotImplementedError

class Average(AggregateFn):
    def aggregate(values: List[VALID_AGGREGATE_TYPE]):
        return sum(values)/len(values)

class Max(AggregateFn):
    def aggregate(values: List[VALID_AGGREGATE_TYPE]):
        return max(values)

class WeightedAverage(AggregateFn):
    #TODO: figure out how to store weights
    def aggregate(values: List[VALID_AGGREGATE_TYPE]):
        #TODO: compute weighted average

#########################################################
# Results Preprocessors
#########################################################

AggregateResultsPreprocessor(ResultsPreprocessor):
    def __init__(aggregation_fn: AggregateFn, keys: Optional[List[str] = None):
        self.keys = keys
        self.aggregation_fn = aggregation_fn

    def preprocess(self, results: List[Dict]) -> List[Dict]:
        if self.keys is None:
             # TODO: find all keys of type VALID_AGGREGATE_TYPE
        for key in self.keys:
             values = [result[key] for result in results]
             aggregated_results = self.aggregation_fn(values)
             # TODO: do something with the result

AverageResultsPreprocessor(AggregateResultsPreprocessor):
    def __init__(keys: Optional[List[str]):
        super.__init(Average(), keys)

MaxResultsPreprocessor(AggregateResultsPreprocessor):
    def __init__(keys: Optional[List[str]):
        super.__init(Max(), keys)

WeightedAverageResultsPreprocessor(AggregateResultsPreprocessor):
    def __init__(keys: Optional[List[str],  str: weight_key):
        super.__init(keys, WeightedAverage(weight_key))
```

The user can then define their aggregation logic with:

```python
average_all_preprocessor = AverageResultsPreprocessor()
max_time_preprocessor = MaxResultsPreprocessor(["time"])
weighted_average_loss_preprocessor = WeightedAverageResultsPreprocessor(["loss"], "batch_size")

results_preprocessors=[average_all_preprocessor, max_time_preprocessor, weighted_average_loss_preprocessor]
            
trainer.run(train_func, results_preprocessors=results_preprocessors)
```


Thoughts?

Reply:
We could just use `__call__` instead of `aggregate` in this example, right? I like this idea - operating on each metric separately, while less flexible, seems more natural when defining aggregation functions.

Reply:
@matthewdeng This idea sounds great to me! Thank you for the suggestion! 

For `WeightedAverageResultsPreprocessor`, can we allow `weight_key` to be a list as well? Different keys may use different weights to average, e.g. the input can be `keys = [key1, key2, ...]` and `weight_key = [weight1, weight, ...]`. When `weight_key = [one_key]`, we broadcast this weight to all keys in the key list. 

Also any consensus on where to store the aggregated metrics?

Reply:
> For `WeightedAverageResultsPreprocessor`, can we allow `weight_key` to be a list as well? 

I don't think we need to support this with the current implementation since the user can just instantiate multiple `WeightedAverageResultsPreprocessor`s instead (one per desired weight).

Reply:
Yes. Thanks for the confirmation! I realized this. The latest commit implements the weighted average for one key 😄